core/include/cubos/core/memory/unordered_bimap.hpp

```diff
@@ -0,0 +1,164 @@
+/// @file
+/// @brief Class @ref cubos::core::memory::UnorderedBimap.
+/// @ingroup core-memory
+
+#pragma once
+
+#include <unordered_map>
+
+namespace cubos::core::memory
+{
+    /// @brief A bidirectional hash table.
+    /// @tparam L Left type.
+    /// @tparam R Right type.
+    /// @tparam LHash Hash functor type for @p L.
+    /// @tparam RHash Hash functor type for @p R.
+    /// @ingroup core-memory
+    template <typename L, typename R, typename LHash = std::hash<L>, typename RHash = std::hash<R>>
+    class UnorderedBimap final
+    {
+    public:
+        using Iterator = typename std::unordered_map<L, R, LHash>::const_iterator;
+
+        /// @brief Adds a new entry to the map.
+        /// @note If any of the values already exists, the old entries with them are removed.
+        /// @param left Left value.
+        /// @param right Right value.
+        void add(L left, R right)
+        {
+            auto leftIt = mLeftToRight.find(left);
+            if (leftIt != mLeftToRight.end())
+            {
+                mRightToLeft.erase(leftIt->second);
+            }
+
+            auto rightIt = mRightToLeft.find(right);
+            if (rightIt != mRightToLeft.end())
+            {
+                mLeftToRight.erase(rightIt->second);
+            }
```

Comment:
doesn't insert or assign already cover this?

Reply:
Nope, this handles situations like having (1, a), (2, b) and then inserting (1, b).

 If you had only insert_and_emplace, the result would be:
 
## Initial state

### L to R

(1, a)
(2, b)

### R to L

(a, 1)
(b, 2)

## Final state

### L to R

(1, b)
(2, b) <- should have been removed

### R to L

(a, 1) <- should have been removed
(b, 1)